// lavacoffee/dist/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IpBlock = exports.RoutePlanner = exports.PlayerVoiceStates = exports.PlayerStates = exports.LoopMode = exports.EventTypes = exports.OpIncoming = exports.OpCodes = void 0;
var OpCodes;
(function (OpCodes) {
    OpCodes["VoiceUpdate"] = "voiceUpdate";
    OpCodes["Play"] = "play";
    OpCodes["Stop"] = "stop";
    OpCodes["Pause"] = "pause";
    OpCodes["Seek"] = "seek";
    OpCodes["Volume"] = "volume";
    OpCodes["Filters"] = "filters";
    OpCodes["Destroy"] = "destroy";
    OpCodes["ConfigResume"] = "configureResuming";
})(OpCodes = exports.OpCodes || (exports.OpCodes = {}));
var OpIncoming;
(function (OpIncoming) {
    OpIncoming["PlayerUpdate"] = "playerUpdate";
    OpIncoming["Stats"] = "stats";
    OpIncoming["Event"] = "event";
})(OpIncoming = exports.OpIncoming || (exports.OpIncoming = {}));
var EventTypes;
(function (EventTypes) {
    EventTypes["TrackStart"] = "TrackStartEvent";
    EventTypes["TrackEnd"] = "TrackEndEvent";
    EventTypes["TrackException"] = "TrackExceptionEvent";
    EventTypes["TrackStuck"] = "TrackStuckEvent";
    EventTypes["WebSocketClosed"] = "WebSocketClosedEvent";
})(EventTypes = exports.EventTypes || (exports.EventTypes = {}));
var LoopMode;
(function (LoopMode) {
    LoopMode[LoopMode["None"] = 0] = "None";
    LoopMode[LoopMode["Track"] = 1] = "Track";
    LoopMode[LoopMode["Queue"] = 2] = "Queue";
})(LoopMode = exports.LoopMode || (exports.LoopMode = {}));
var PlayerStates;
(function (PlayerStates) {
    PlayerStates[PlayerStates["Playing"] = 0] = "Playing";
    PlayerStates[PlayerStates["Paused"] = 1] = "Paused";
    PlayerStates[PlayerStates["Destroyed"] = 2] = "Destroyed";
})(PlayerStates = exports.PlayerStates || (exports.PlayerStates = {}));
var PlayerVoiceStates;
(function (PlayerVoiceStates) {
    PlayerVoiceStates[PlayerVoiceStates["Connecting"] = 0] = "Connecting";
    PlayerVoiceStates[PlayerVoiceStates["Connected"] = 1] = "Connected";
    PlayerVoiceStates[PlayerVoiceStates["Disconnecting"] = 2] = "Disconnecting";
    PlayerVoiceStates[PlayerVoiceStates["Disconnected"] = 3] = "Disconnected";
})(PlayerVoiceStates = exports.PlayerVoiceStates || (exports.PlayerVoiceStates = {}));
var RoutePlanner;
(function (RoutePlanner) {
    RoutePlanner["Balancing"] = "BalancingIpRoutePlanner";
    RoutePlanner["Rotating"] = "RotatingIpRoutePlanner";
    RoutePlanner["Nano"] = "NanoIpRoutePlanner";
    RoutePlanner["RotatingNano"] = "RotatingNanoIpRoutePlanner";
})(RoutePlanner = exports.RoutePlanner || (exports.RoutePlanner = {}));
var IpBlock;
(function (IpBlock) {
    IpBlock["Ipv4"] = "Inet4Address";
    IpBlock["Ipv6"] = "Inet6Address";
})(IpBlock = exports.IpBlock || (exports.IpBlock = {}));
